File: scripts/evalCopilot.ts
/**
 * Live smoke for the builder copilot: one real conversation turn against the
 * local CLIProxyAPI. Asks for a section-instruction tightening and asserts the
 * turn produced at least one applied edit op and a coherent reply.
 * Exit codes match scripts/eval.ts: 0 ok, 1 failed, 2 proxy unreachable, 3 auth.
 */
import { join } from "node:path";
import { pathToFileURL } from "node:url";
import OpenAI from "openai";
import { openDb, BlueprintContentSchema, formatSqlResult, readWarehouseTables, runWarehouseSql, type RunoffDb } from "@runoff/core";
import {
  makeLlmClient,
  copilotTurn,
  MODEL,
  type CatalogFamily,
  type CopilotContext,
  type EngineFile,
  type FamilyInfo,
} from "@runoff/engine";
import { seedDatabase } from "./seed.js";

const BLUEPRINT_NAME = "Quarterly Performance Report";

function dbPath(): string {
  return process.env.RUNOFF_DB ?? "data/runoff.db";
}

function filesDir(): string {
  return process.env.RUNOFF_FILES_DIR ?? "data/files";
}

function baseURL(): string {
  return process.env.OPENAI_BASE_URL ?? "http://localhost:8317/v1";
}

/**
 * Probe the proxy before spending a full run. Resolves on success; otherwise
 * prints a SKIPPED line and exits with a distinct code (2 unreachable, 3 auth).
 */
async function preflight(client: OpenAI): Promise<void> {
  try {
    await client.models.list();
  } catch (err) {
    if (
      err instanceof OpenAI.APIConnectionError ||
      (err as { code?: string })?.code === "ECONNREFUSED"
    ) {
      console.error(`EVAL SKIPPED — CLIProxyAPI unreachable at ${baseURL()}`);
      process.exit(2);
    }
    const status = (err as { status?: number })?.status;
    if (err instanceof OpenAI.AuthenticationError || status === 401 || status === 403) {
      console.error("EVAL SKIPPED — proxy requires a valid OPENAI_API_KEY (set it and re-run pnpm eval)");
      process.exit(3);
    }
    // Any other failure (e.g. proxy up but /models unimplemented): re-probe with
    // a tiny chat completion, which exercises the exact path the run uses.
    try {
      await client.chat.completions.create({
        model: MODEL,
        max_tokens: 1,
        messages: [{ role: "user", content: "ping" }],
      });
    } catch (err2) {
      if (
        err2 instanceof OpenAI.APIConnectionError ||
        (err2 as { code?: string })?.code === "ECONNREFUSED"
      ) {
        console.error(`EVAL SKIPPED — CLIProxyAPI unreachable at ${baseURL()}`);
        process.exit(2);
      }
      const s2 = (err2 as { status?: number })?.status;
      if (err2 instanceof OpenAI.AuthenticationError || s2 === 401 || s2 === 403) {
        console.error("EVAL SKIPPED — proxy requires a valid OPENAI_API_KEY (set it and re-run pnpm eval)");
        process.exit(3);
      }
      console.error(`EVAL SKIPPED — proxy probe failed: ${err2 instanceof Error ? err2.message : String(err2)}`);
      process.exit(2);
    }
  }
}

/** Load the seeded blueprint (seeding first if absent). */
async function loadBlueprint(db: RunoffDb): Promise<{ blueprintId: string; rev: number; content: ReturnType<typeof BlueprintContentSchema.parse> }> {
  let row = db.sqlite
    .prepare("SELECT id, current_rev AS rev FROM blueprints WHERE name = ?")
    .get(BLUEPRINT_NAME) as { id: string; rev: number } | undefined;
  if (!row) {
    console.log(`No "${BLUEPRINT_NAME}" found — seeding first…`);
    const { blueprintId } = await seedDatabase(db);
    row = db.sqlite.prepare("SELECT id, current_rev AS rev FROM blueprints WHERE id = ?").get(blueprintId) as {
      id: string;
      rev: number;
    };
  }

  const revRow = db.sqlite
    .prepare("SELECT content FROM blueprint_revisions WHERE blueprint_id = ? AND rev = ?")
    .get(row.id, row.rev) as { content: string } | undefined;
  if (!revRow) throw new Error(`blueprint revision not found: ${row.id}@${row.rev}`);
  const content = BlueprintContentSchema.parse(JSON.parse(revRow.content));

  return { blueprintId: row.id, rev: row.rev, content };
}

/**
 * Build the Task-10 CopilotContext for the seeded blueprint, mirroring
 * `buildCopilotContext` in apps/web/lib/queries.ts: families (with bound /
 * filedPeriods / hasLiveFile), defaultFiles (bound families → live file, id =
 * family id), and periodFiles (every filed periodic row of bound families). Run
 * history / goldens are stubbed empty — the eval only exercises source access,
 * edits, and saveMemory.
 */
export function buildEvalContext(db: RunoffDb, blueprintId: string): CopilotContext {
  const dir = filesDir();
  const projectId =
    (db.sqlite.prepare("SELECT project_id AS projectId FROM blueprints WHERE id = ?").get(blueprintId) as
      | { projectId: string }
      | undefined)?.projectId ?? "";
  const boundSet = new Set(
    (db.sqlite
      .prepare("SELECT family_id AS familyId FROM blueprint_families WHERE blueprint_id = ?")
      .all(blueprintId) as { familyId: string }[]).map((r) => r.familyId),
  );
  const famRows = db.sqlite
    .prepare("SELECT id, key, label, kind, granularity FROM source_families WHERE project_id = ? ORDER BY key")
    .all(projectId) as {
    id: string;
    key: string;
    label: string;
    kind: "periodic" | "constant";
    granularity: FamilyInfo["granularity"];
  }[];
  const periodsStmt = db.sqlite.prepare(
    "SELECT period FROM sources WHERE family_id = ? AND status='filed' AND period IS NOT NULL ORDER BY period",
  );
  const liveStmt = db.sqlite.prepare(
    "SELECT 1 FROM sources WHERE family_id = ? AND status='filed' AND period IS NULL LIMIT 1",
  );
  const families: FamilyInfo[] = famRows.map((f) => ({
    id: f.id,
    key: f.key,
    label: f.label,
    kind: f.kind,
    granularity: f.granularity,
    filedPeriods:
      f.kind === "constant" ? [] : (periodsStmt.all(f.id) as { period: string }[]).map((r) => r.period),
    hasLiveFile: f.kind === "constant" ? !!liveStmt.get(f.id) : false,
    bound: boundSet.has(f.id),
  }));

  const constantSlot = db.sqlite.prepare(
    "SELECT mime, stored_filename AS storedFilename FROM sources WHERE family_id = ? AND status='filed' AND period IS NULL",
  );
  const latestSlot = db.sqlite.prepare(
    "SELECT mime, stored_filename AS storedFilename FROM sources WHERE family_id = ? AND status='filed' AND period IS NOT NULL ORDER BY period DESC LIMIT 1",
  );
  const defaultFiles: EngineFile[] = [];
  for (const f of families) {
    if (!f.bound) continue;
    const r = (f.kind === "constant" ? constantSlot : latestSlot).get(f.id) as
      | { mime: string; storedFilename: string }
      | undefined;
    if (!r) continue;
    defaultFiles.push({ id: f.id, name: f.label, mime: r.mime, path: join(dir, r.storedFilename) });
  }

  const periodRowsStmt = db.sqlite.prepare(
    "SELECT period, mime, stored_filename AS storedFilename FROM sources WHERE family_id = ? AND status='filed' AND period IS NOT NULL ORDER BY period",
  );
  const periodFiles: CopilotContext["periodFiles"] = [];
  for (const f of families) {
    if (!f.bound || f.kind !== "periodic") continue;
    for (const r of periodRowsStmt.all(f.id) as { period: string; mime: string; storedFilename: string }[]) {
      periodFiles.push({
        familyId: f.id,
        period: r.period,
        file: { id: f.id, name: f.label, mime: r.mime, path: join(dir, r.storedFilename) },
      });
    }
  }

  const catalog: CatalogFamily[] = families.map((f) => {
    const tables = readWarehouseTables(projectId, f.key);
    return {
      id: f.id,
      key: f.key,
      label: f.label,
      kind: f.kind,
      granularity: f.granularity,
      queryable: tables.length > 0,
      tables,
      filedPeriods: f.filedPeriods,
    };
  });

  return {
    families,
    defaultFiles,
    periodFiles,
    catalog,
    runSql: (sql: string) => formatSqlResult(runWarehouseSql(projectId, sql)),
    listRuns: () => [],
    getRunSection: () => null,
    listGoldens: () => [],
    getGolden: () => null,
    saveMemory: () => "mem_eval",
  };
}

async function main(): Promise<void> {
  const client = makeLlmClient();
  await preflight(client);
  const db = openDb(dbPath());
  try {
    const { blueprintId, content } = await loadBlueprint(db);
    let editOps = 0;
    const ctx: CopilotContext = buildEvalContext(db, blueprintId);
    const res = await copilotTurn({
      client,
      draft: content,
      selectedKey: content.sections[0]?.key ?? null,
      message:
        "Tighten the first auto section's instruction: demand exactly three sentences and require the headline spend figure to be cited. Apply the edit.",
      thread: [],
      memories: [],
      ctx,
      io: {
        emit(e) {
          if (e.type === "tool_activity") console.log(`[tool] ${e.label}`);
          if (e.type === "edit") { editOps++; console.log(`[edit] ${e.op.type}`); }
          if (e.type === "text_delta") process.stdout.write(".");
        },
      },
    });
    process.stdout.write("\n");
    const failures: string[] = [];
    if (editOps < 1) failures.push("no edit op was applied");
    if (res.reply.trim().length < 10) failures.push(`reply too short: "${res.reply}"`);
    if (failures.length) {
      console.error(`\nCOPILOT EVAL FAILED — ${failures.join("; ")}`);
      process.exit(1);
    }
    console.log(`\nCOPILOT EVAL OK — ${editOps} edit op(s), ${res.actions.length} action(s), reply ${res.reply.length} chars`);
  } finally {
    db.sqlite.close();
  }
}

// Only run the live copilot eval when this module is the direct entrypoint.
// scripts/evalSql.ts imports buildEvalContext from here; the guard keeps that
// import side-effect-free (no unintended second live run, no coupled exit code).
if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  main().catch((err) => {
    console.error(`\nCOPILOT EVAL FAILED — ${err instanceof Error ? err.message : String(err)}`);
    process.exit(1);
  });
}

File: backend/evals/eval_copilot.py
"""Live smoke for the builder copilot — Python twin of scripts/evalCopilot.ts.

One real conversation turn against the local CLIProxyAPI: ask for a
section-instruction tightening on the seeded "Quarterly Performance Report"
blueprint and assert the turn produced at least one applied edit op and a
coherent reply. Unlike the TS twin (which stubs the golden callbacks), this runs
the turn against the REAL Python services — caches + context are built through
resolve_golden / render_golden_for_prompt / scaffold_digest_for /
build_copilot_context, exactly as the copilot route does — over the
already-seeded dev DB (run `pnpm seed` first if the blueprint is absent).

Exit codes match evalCopilot.ts: 0 ok, 1 failed, 2 proxy unreachable, 3 auth.
Plain script, no pytest.
"""

import os
import sys
from pathlib import Path

# Running as a script file puts only backend/evals/ on sys.path; add backend/ so
# the runoff_api package (backend/runoff_api) imports without an install step.
sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import openai  # noqa: E402

from runoff_api.core.db import open_db  # noqa: E402
from runoff_api.core.types.blueprint import BlueprintContent  # noqa: E402
from runoff_api.engine.copilot import copilot_turn  # noqa: E402
from runoff_api.engine.llm import make_llm_client  # noqa: E402
from runoff_api.engine.prompts import MODEL  # noqa: E402
from runoff_api.services.golden_binding import boundness_line, render_golden_for_prompt  # noqa: E402
from runoff_api.services.goldens import list_goldens, resolve_golden, scaffold_digest_for  # noqa: E402
from runoff_api.services.queries import build_copilot_context  # noqa: E402

BLUEPRINT_NAME = "Quarterly Performance Report"


def db_path() -> str:
    return os.environ.get("RUNOFF_DB", "data/runoff.db")


def base_url() -> str:
    return os.environ.get("OPENAI_BASE_URL", "http://localhost:8317/v1")


def preflight(client: openai.OpenAI) -> None:
    """Probe the proxy before spending a full run. Mirrors evalCopilot.ts:
    prints a SKIPPED line and exits 2 (unreachable) / 3 (auth) on failure."""
    try:
        client.models.list()
    except Exception as err:  # noqa: BLE001
        if isinstance(err, openai.APIConnectionError) or getattr(err, "code", None) == "ECONNREFUSED":
            print(f"EVAL SKIPPED — CLIProxyAPI unreachable at {base_url()}", file=sys.stderr)
            sys.exit(2)
        status = getattr(err, "status_code", None) or getattr(err, "status", None)
        if isinstance(err, openai.AuthenticationError) or status in (401, 403):
            print(
                "EVAL SKIPPED — proxy requires a valid OPENAI_API_KEY (set it and re-run pnpm backend:eval:copilot)",  # noqa: E501
                file=sys.stderr,
            )
            sys.exit(3)
        # Any other failure (e.g. proxy up but /models unimplemented): re-probe with
        # a tiny chat completion, which exercises the exact path the run uses.
        try:
            client.chat.completions.create(
                model=MODEL, max_completion_tokens=1, messages=[{"role": "user", "content": "ping"}]
            )
        except Exception as err2:  # noqa: BLE001
            if isinstance(err2, openai.APIConnectionError) or getattr(err2, "code", None) == "ECONNREFUSED":
                print(f"EVAL SKIPPED — CLIProxyAPI unreachable at {base_url()}", file=sys.stderr)
                sys.exit(2)
            s2 = getattr(err2, "status_code", None) or getattr(err2, "status", None)
            if isinstance(err2, openai.AuthenticationError) or s2 in (401, 403):
                print(
                    "EVAL SKIPPED — proxy requires a valid OPENAI_API_KEY (set it and re-run pnpm backend:eval:copilot)",  # noqa: E501
                    file=sys.stderr,
                )
                sys.exit(3)
            print(f"EVAL SKIPPED — proxy probe failed: {err2}", file=sys.stderr)
            sys.exit(2)


def load_blueprint(db) -> tuple[str, dict]:
    """Load the seeded blueprint's current-revision content (mirrors loadBlueprint)."""
    row = db.execute(
        "SELECT id, current_rev AS rev FROM blueprints WHERE name = ?", (BLUEPRINT_NAME,)
    ).fetchone()
    if row is None:
        print(f'COPILOT EVAL FAILED — no "{BLUEPRINT_NAME}" blueprint (run pnpm seed first)', file=sys.stderr)
        sys.exit(1)
    rev_row = db.execute(
        "SELECT content FROM blueprint_revisions WHERE blueprint_id = ? AND rev = ?",
        (row["id"], row["rev"]),
    ).fetchone()
    if rev_row is None:
        raise RuntimeError(f"blueprint revision not found: {row['id']}@{row['rev']}")
    import json

    content = BlueprintContent.model_validate(json.loads(rev_row["content"]))
    return row["id"], content.model_dump(by_alias=True, exclude_unset=True)


def build_context(db, blueprint_id: str) -> dict:
    """Build caches + context through the real services, exactly as the copilot route does."""
    golden_cache: dict[str, dict] = {}
    scaffold_cache: dict[str, str] = {}
    for g in list_goldens(db, blueprint_id):
        resolved = resolve_golden(db, g["id"])
        if resolved is None:
            continue
        golden_cache[g["id"]] = {
            "description": f"{resolved['label']} — {boundness_line(resolved['inventory'])}",
            "text": render_golden_for_prompt(resolved),
        }
        scaffold_cache[g["id"]] = scaffold_digest_for(resolved)
    return build_copilot_context(db, blueprint_id, golden_cache, scaffold_cache)


def main() -> None:
    client = make_llm_client()
    preflight(client)
    db = open_db(db_path())
    try:
        blueprint_id, content = load_blueprint(db)
        ctx = build_context(db, blueprint_id)
        edit_ops = 0

        class Io:
            def emit(self, e: dict) -> None:
                nonlocal edit_ops
                if e["type"] == "tool_activity":
                    print(f"[tool] {e['label']}")
                elif e["type"] == "edit":
                    edit_ops += 1
                    print(f"[edit] {e['op']['type']}")
                elif e["type"] == "text_delta":
                    sys.stdout.write(".")
                    sys.stdout.flush()

        res = copilot_turn(
            client=client,
            draft=content,
            selected_key=(content["sections"][0]["key"] if content.get("sections") else None),
            message=(
                "Tighten the first auto section's instruction: demand exactly three sentences "
                "and require the headline spend figure to be cited. Apply the edit."
            ),
            thread=[],
            memories=[],
            ctx=ctx,
            io=Io(),
        )
        sys.stdout.write("\n")
        failures: list[str] = []
        if edit_ops < 1:
            failures.append("no edit op was applied")
        if len(res["reply"].strip()) < 10:
            failures.append(f'reply too short: "{res["reply"]}"')
        if failures:
            print(f"\nCOPILOT EVAL FAILED — {'; '.join(failures)}", file=sys.stderr)
            sys.exit(1)
        print(
            f"\nCOPILOT EVAL OK — {edit_ops} edit op(s), {len(res['actions'])} action(s), "
            f"reply {len(res['reply'])} chars"
        )
    finally:
        db.close()


if __name__ == "__main__":
    try:
        main()
    except SystemExit:
        raise
    except Exception as err:  # noqa: BLE001
        print(f"\nCOPILOT EVAL FAILED — {err}", file=sys.stderr)
        sys.exit(1)
